src/main/java/org/prebid/server/bidder/huaweiads/HuaweiAdsBidder.java

```diff
@@ -0,0 +1,1294 @@
+        } else if (numMainImage > 1) {
+            detailedCreativeTypeList.add("904");
+        } else if (numMainImage == 1) {
+            detailedCreativeTypeList.add("901");
+        } else {
+            detailedCreativeTypeList.add("913");
+            detailedCreativeTypeList.add("914");
+        }
+
+        adSlot30Builder.detailedCreativeTypeList(detailedCreativeTypeList);
+
+        return adSlot30Builder.build();
+    }
+
+    private AdSlot30 getVideoFormat(AdSlot30 adslot30, int adtype, Imp openRTBImp) {
+        AdSlot30.AdSlot30Builder adSlot30Builder = adslot30.toBuilder();
+        adSlot30Builder.w(openRTBImp.getVideo().getW());
+        adSlot30Builder.h(openRTBImp.getVideo().getH());
+
+        if (adtype == AdsType.ROLL.getCode()) {
+            if (openRTBImp.getVideo().getMaxduration() == null || openRTBImp.getVideo().getMaxduration() == 0) {
+                throw new PreBidException(
+                        "Extract openrtb video failed: MaxDuration is empty when huaweiads adtype is roll.");
+            }
+            adSlot30Builder.totalDuration(openRTBImp.getVideo().getMaxduration());
+        }
+        return adSlot30Builder.build();
+    }
+
+    private HuaweiAdsRequest getHuaweiAdsRequest(HuaweiAdsRequest request, BidRequest openRTBRequest) {
+        HuaweiAdsRequest.HuaweiAdsRequestBuilder huaweiAdsRequestBuilder = request.toBuilder();
+        huaweiAdsRequestBuilder.version(HuaweiAdsConstants.API_VERSION);
+
+        huaweiAdsRequestBuilder.app(getAppInfo(openRTBRequest));
+        huaweiAdsRequestBuilder.device(getDeviceInfo(openRTBRequest));
+        huaweiAdsRequestBuilder.network(getNetworkInfo(openRTBRequest));
+        huaweiAdsRequestBuilder.regs(getRegsInfo(openRTBRequest));
+        huaweiAdsRequestBuilder.geo(getGeoInfo(openRTBRequest));
+        huaweiAdsRequestBuilder.consent(getConsentInfo(openRTBRequest));
+
+        return huaweiAdsRequestBuilder.build();
+    }
+
+    private App getAppInfo(BidRequest openRTBRequest) {
+        final App.AppBuilder app = App.builder();
+        if (openRTBRequest.getApp() != null) {
+            if (openRTBRequest.getApp().getVer() != null && !openRTBRequest.getApp().getVer().isEmpty()) {
+                app.version(openRTBRequest.getApp().getVer());
+            }
+            if (openRTBRequest.getApp().getName() != null && !openRTBRequest.getApp().getName().isEmpty()) {
+                app.name(openRTBRequest.getApp().getName());
+            }
+
+            // bundle cannot be empty, we need package name.
+            if (openRTBRequest.getApp().getBundle() != null && !openRTBRequest.getApp().getBundle().isEmpty()) {
+                app.pkgname(getFinalPackageName(openRTBRequest.getApp().getBundle()));
+            } else {
+                throw new PreBidException("generate HuaweiAds AppInfo failed: openrtb BidRequest.App.Bundle is empty.");
+            }
+
+            if (openRTBRequest.getApp().getContent() != null
```

Comment:
You have multiple calls like `openRTBRequest.getApp()` it can be extracted to separate variable.
Pls check similar occurences